### ISSUE TITLE
`normalize_tags` is quite expensive in aggregate.

### ISSUE THREAD

**Issue by ptolts** [Shopify/statsd-instrument#357]
I've been profiling application code and I've noticed that `normalize_tags` is quite expensive when statsd is used at scale. The profile below represents a significant overhead for the profile in question.

`statsd-instrument 3.5.12`

<img width="1148" alt="image" src="https://github.com/Shopify/statsd-instrument/assets/723514/d61672a0-ccdb-439b-96bd-9f7ff1317557">

```
   def normalize_tags(tags)
        return [] unless tags

        tags = tags.map { |k, v| "#{k}:#{v}" } if tags.is_a?(Hash)

        # Fast path when no string replacement is needed
        return tags unless tags.any? { |tag| /[|,]/.match?(tag) }

        tags.map { |tag| tag.tr("|,", "") }
      end
```

https://github.com/Shopify/statsd-instrument/blob/e39f8ec3a3c18f6b7625b5686e94abea1ceb1301/lib/statsd/instrument/datagram_builder.rb#L70-L79

**Comment by casperisfine** [comment]
Yeah, it's in big part caused by how liberal it is in the arguments it accepts. If it was stricter, we could optimize a bit better.

I'll still see if I can pull something

**Comment by wvanbergen** [comment]
It's been ages, but I think there is a way to call the statsd instrumentation functions that bypasses most of the normalization and cleanup code, specifically for hot code paths. I remember porting a bunch of the hot paths in the monolith.

**Comment by casperisfine** [comment]
O, #358 made the datagram builder ~1.4x faster (I kept the normalization).

Overall `benchmark/send-metrics-to-dev-null-log` is 1.15x faster, but it uses fairly few tags so might be a bit more in real world.

I don't think there's much more we can do, so if after this change it's still a significant part of your request time, I'd suggest looking at how many metrics you emit and consider if they're all really useful.
